### PR TITLE
go.mod: upgrade opentelemetry deps

### PR DESCRIPTION
**Pull request by willbeason**
Downstream users of etcd experience build issues when using dependencies
which require more recent (incompatible) versions of opentelemetry. This
commit upgrades the dependencies so that downstream users stop
experiencing these issues.

This will resolve:
- https://github.com/open-policy-agent/frameworks/issues/210
- https://github.com/kubernetes/component-base/issues/5

### CODE REVIEW

**Comment by serathius** [comment]
FYI this is third PR that tried to upgrade opentelemetry, however others didn't pass tests and seem abandoned. https://github.com/etcd-io/etcd/pull/13723 https://github.com/etcd-io/etcd/pull/13808.

**Comment by willbeason** [comment]
> FYI this is third PR that tried to upgrade opentelemetry, however others didn't pass tests and seem abandoned. https://github.com/etcd-io/etcd/pull/13723 https://github.com/etcd-io/etcd/pull/13808. 

Ok. I'm having trouble seeing the failed test results. I only see:

"Error: The log was not found. It may have been deleted based on retention settings."

**Comment by willbeason** [comment]
Is there a way to run the workflow locally?

**Comment by serathius** [comment]
Error in Static analysis:
```
FAIL: inconsistent versions for depencency: google.golang.org/grpc
  - go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.26.1 from: go.etcd.io/etcd/server/v3
  - google.golang.org/grpc@v1.41.0 from: go.etcd.io/etcd/api/v3
  - google.golang.org/grpc@v1.41.0 from: go.etcd.io/etcd/client/v3
  - google.golang.org/grpc@v1.41.0 from: go.etcd.io/etcd/etcdctl/v3
FAIL: inconsistent versions for depencency: google.golang.org/protobuf
FAIL: inconsistent dependencies
  - google.golang.org/grpc@v1.41.0 from: go.etcd.io/etcd/pkg/v3
FAIL: 'dep' failed at Tue Apr  5 18:27:09 UTC 2022
```

**Comment by willbeason** [comment]
Thanks! I'll take a look tomorrow and correct that

**Comment by willbeason** [comment]
For future people: you can run static analysis yourself. For the above failure (inconsistent dependencies), for example, I can use this command:
```
PASSES='dep' ./scripts/test.sh
```

This isn't guaranteed to catch all failures as there could be architecture-specific failures, but it will likely catch most of them.

**Comment by willbeason** [comment]
I see - the main intent of these failures is that:

- this repository contains multiple Go modules
- the 'dep' check ensures that every external dependency has exactly the same version for all of these modules

**Comment by willbeason** [comment]
Ok - new failure this time, so making progress.

Here is the failure: https://gist.github.com/willbeason/c774057148da7dc69fb1bf0fbc5d166c

This failure can be recreated with the command:
```
PASSES='bom' ./scripts/test.sh
```

It looks like a check that ensures the various tools used by scripts are the ones installed by a default invocation of `go install`. In this case, `bill-of-materials`.

**Comment by willbeason** [comment]
The solution here was to run this script which automatically fixes these issues:

```
./scripts/fix.sh
```

**Comment by willbeason** [comment]
That's odd - the only change in the code is running `./scripts/fix.sh` (I verified there were no spurious changes), but now Semaphore reports that the build fails for i386 architecture, whereas for the previous commit it passed. The only changes are to bill-of-materials files and to go.sum. Could it be a flake, or have I missed something?

**Comment by willbeason** [comment]
Ah, I see from the failure that it is a flake:

```
Step 1/14 : FROM ubuntu:21.10
toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
make[1]: *** [build-docker-test] Error 1
make[1]: Leaving directory `/home/runner/etcd'
make: *** [ensure-docker-test-image-exists] Error 2
```

**Comment by serathius** [comment]
Yep, it's flaky. https://github.com/etcd-io/etcd/issues/13448

**Comment by willbeason** [comment]
Hooray, static analysis passed!

**Comment by serathius** [comment]
@willbeason appreciate any feedback on how we could make contributions easier :P.

**Comment by serathius** [comment]
Created https://github.com/etcd-io/etcd/issues/13896, feel free to add more.

**Comment by willbeason** [comment]
Thanks!

I'm really hoping that this PR doesn't run into the same issues that the previous PRs ran into. The deps here are higher versions than the others, but the related issues haven't been closed, so fingers crossed.

**Comment by willbeason** [comment]
Drat! Integration tests failing with timeouts like previous PRs.

**Comment by willbeason** [comment]
Heads up: I'll be intermittently available tomorrow, and I'm OOO Friday. I'll be back Monday. I'll still be on for a few more hours today.

Looking at first failure: https://gist.github.com/willbeason/e6b372803acbf58d16bc27a533d338ac

So for the TestV3StorageQuotaApply, there's a goroutine leak before the `TestV3StorageQuotaApply` test even starts. That should help narrow down when the goroutine is leaked.

**Comment by willbeason** [comment]
That's odd - running the test in isolation locally passes:
```
tests$ go test ./integration/ --run TestV3StorageQuotaApply --race --count=10
ok  	go.etcd.io/etcd/tests/v3/integration	14.924s
```
(note that since I ran the test 10x, the runtime is 10x longer than usual)

So this means that a test which runs _before_ TestV3StorageQuotaApply causes the problem

**Comment by willbeason** [comment]
@serathius I see that while many tests in `tests/integration` run with `integration2.BeforeTest(t)` as the first call, not all do. Is that intentional? Since the goroutine is leaked before `TestV3StorageQuotaApply` is run, then another test must be leaking the goroutine (since we know TestMain does not). Tests which call `integration2.BeforeTest(t)` check no goroutines are leaked as a cleanup step, so IIUC then one of the few tests which does not call `integration2.BeforeTest(t)` must be leaking the goroutine.

Do I have this right?

**Review by ptabor — APPROVED** [review]

**Comment by serathius** [comment]
@willbeason yes, we have a lot of flaky tests and have effort to track and fix them. Please ignore the test failures if they are flakes. I will rerun the test if needed.

Please fix the conflict and rebase the PR.

**Comment by ptabor** [comment]
```
go: golang.org/x/crypto@v0.0.0-20210513164829-c07d[79](https://github.com/etcd-io/etcd/runs/5956315508?check_suite_focus=true#step:5:79)3c2f9a: missing go.sum entry; to add it:
	go mod download golang.org/x/crypto
must be run from 'go.etcd.io/etcd/v3' module directory
```

**Comment by willbeason** [comment]
Updated, and manually verified locally that "dep" and "bom" pass

File: client/pkg/testutil/leak.go [hunk 1]
```diff
@@ -99,7 +99,7 @@ func CheckAfterTest(d time.Duration) error {
 // If you execute RegisterLeakDetection, you don't need to explicitly register AfterTest.
 func RegisterLeakDetection(t TB) {
 	if err := CheckAfterTest(10 * time.Millisecond); err != nil {
-		t.Skip("Found leaked goroutined BEFORE test", err)
+		t.Skip("Found leaked goroutine BEFORE test", err)
```
Comment (ahrtr) [inline client/pkg/testutil/leak.go:102]: goroutines? I believe previous the letter 'd' is just a typo of 's'.

**Review by ahrtr — APPROVED** [review]
LGTM with a nit

**Comment by willbeason** [comment]
@ahrtr Fixed! I don't think I have permissions to merge.

**Comment by ptabor** [comment]
Please squash the commits and I will merge as soon as the tests are green.

File: client/pkg/testutil/leak.go (hunk 1 above)
Comment (willbeason) [reply client/pkg/testutil/leak.go]: Done

**Comment by willbeason** [comment]
@ptabor Ok - squashed commits

**Pull request merged**